Guarantee aligned resources across containers

klueska opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR implements the logic required to resolve the issue outlined in https://github.com/kubernetes/kubernetes/issues/83476. It does this by changing the pod admit handler for the TopologyManager to loop through each container, first calling out to the CPUManager and devicemanager to generate hints and then immediately calling out to perform device allocation in the same iteration of the loop. Glue code around this was changed as well to keep all of the abstractions clean and usable.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #83476

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug in the TopologyManager. Previously, the TopologyManager would only guarantee alignment if container creation was serialized in some way. Alignment is now guaranteed under all scenarios of container creation. 
```

klueska committed:
Split devicemanager Allocate into two functions

Instead of having a single call for Allocate(), we now split this into two
functions Allocate() and UpdatePluginResources().

The semantics split across them:

// Allocate configures and assigns devices to a pod. From the requested
// device resources, Allocate will communicate with the owning device
// plugin to allow setup procedures to take place, and for the device
// plugin to provide runtime settings to use the device (environment
// variables, mount points and device files).
Allocate(pod *v1.Pod) error

// UpdatePluginResources updates node resources based on devices already
// allocated to pods. The node object is provided for the device manager to
// update the node capacity to reflect the currently available devices.
UpdatePluginResources(
    node *schedulernodeinfo.NodeInfo,
    attrs *lifecycle.PodAdmitAttributes) error

As we move to a model in which the TopologyManager is able to ensure
aligned allocations from the CPUManager, devicemanger, and any
other TopologManager HintProviders in the same synchronous loop, we will
need to be able to call Allocate() independently from an
UpdatePluginResources(). This commit makes that possible.

klueska committed:
Add Allocate() call to TopologyManager's HintProvider interface

Having this interface allows us to perform a tight loop of:

    for each container {
        containerHints = {}
        for each provider {
	    containerHints[provider] = provider.GatherHints(container)
        }

        containerHints.MergeAndPublish()

        for each provider {
	    provider.Allocate(container)
        }
    }

With this in place we can now be sure that the hints gathered in one
iteration of the loop always consider the allocations made in the
previous.

klueska committed:
Change CPUManager to implement HintProvider.Allocate()

This change will not work on its own. Higher level code needs to make
sure and call Allocate() before AddContainer is called. This is already
being done in cases when the TopologyManager feature gate is enabled (in
the PodAdmitHandler of the TopologyManager). However, we need to make
sure we add proper logic to call it in cases when the TopologyManager
feature gate is disabled.

klueska committed:
Change devicemanager to implement HintProvider.Allocate()

This change will not work on its own. Higher level code needs to make
sure and call Allocate() before AddContainer is called. This is already
being done in cases when the TopologyManager feature gate is enabled (in
the PodAdmitHandler of the TopologyManager). However, we need to make
sure we add proper logic to call it in cases when the TopologyManager
feature gate is disabled.

klueska committed:
Change GetTopologyPodAdmitHandler() to be more general

GetTopologyPodAdmitHandler() now returns a lifecycle.PodAdmitHandler
type instead of the TopologyManager directly. The handler it returns
is generally responsible for attempting to allocate any resources that
require a pod admission check. When the TopologyManager feature gate
is on, this comes directly from the TopologyManager. When it is off,
we simply attempt the allocations ourselves and fail the admission
on an unexpected error. The higher level kubelet.go feature gate
check will be removed in an upcoming PR.

bg-chun commented:
Are you going to make PR merged in 1.18 rel?

klueska commented:
/hold

klueska commented:
@bg-chun yes that is the plan. There’s still quite a bit of work to do on the PR though, and I will not have much time to work on it anymore between now and the code freeze.

The plan is for @vpickard and @nolancon to pick it up from here and continue developing it.

nolancon commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -206,46 +211,57 @@ func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesRe
+	//			klog.Warningf("[cpumanager] unable to remove init container (pod: %s, container: %s, error: %v)", string(p.UID), initContainer.Name, err)
+	//		}
+	//	}
+	//}
```
@klueska @vpickard would it suffice to do a check for init container in `policy.AddContainer()`  and then remove that container's CPUs from the shared pool? 
So at the bottom of `policy.AddContainer()` we could call out to something like:
``````
func (p *staticPolicy) releaseInitContainerCPUs(s state.State, pod *v1.Pod, container *v1.Container) {
        for _, initContainer := range pod.Spec.InitContainers {
                if container.Name == initContainer.Name {
                        toRelease, _ := s.GetCPUSet(string(pod.UID), container.Name)
                        // Mutate the shared pool, adding released cpus.
                        s.SetDefaultCPUSet(s.GetDefaultCPUSet().Union(toRelease))
                }
        }
}

klueska replied:
I'd need to look closer at this, but is the idea to:
1) Unconditionally allocate CPUs to the container from the pool of available CPUs 
2) Check if the container we just allocated to is an init container
3) if it IS an init container, reset the pool of available CPUs to re-include the CPUs just assigned to the init container (but keep them assigned to the init container in the process).
4) If it is NOT an init container, just return (leaving the CPUs removed from the pool of available CPUs).

I think this would probably work, and it's actually fairly simple to understand.

In the process, it probably makes sense to rename the function to policy.Allocate() instead of policy.AddContainer() since it is (and actually always was) just doing the CPU allocation, and not the actual taskset assignment through the container runtime.

nolancon replied:
Yes, this is the flow I was thinking of. It seems like the simplest solution. I will push the fix. 
I'll also rename the function as mentioned.

ipuustin replied:
I guess the problem in this solution is that the init container will not be running on exclusive CPUs even if it has Guaranteed QoS, so some other container from another Pod might end up using the same CPUs?

derekwaynecarr replied:
if i follow this correctly, it should be fine because init containers are run in sequence.  i will review the remaining commits to see if this works out.

ipuustin commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -206,46 +211,57 @@ func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesRe
+	//}
 
 	// Call down into the policy to assign this container CPUs if required.
-	err := m.policyAddContainer(p, c, containerID)
```
Hmm, is the idea here to delay adding the container to the `containerMap` to the reconciliation loop? It seems that reconciliation loop might not get far enough in the loop to actually perform the Add().

klueska replied:
No, the it will be added to the containerMap as soon as it can be (i.e. at the top `AddContainer()` below).

ipuustin replied:
I asked because I only see `m.containerMap.Add()` being called in the reconciliation loop, but maybe it's abstracted somewhere...

Commit:
CPU Manager - Rename policy.AddContainer() to policy.Allocate()

ipuustin commented on pkg/kubelet/kubelet.go:
```diff
@@ -867,9 +867,9 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
-		klet.admitHandlers.AddPodAdmitHandler(klet.containerManager.GetTopologyPodAdmitHandler())
-	}
+
+	klet.admitHandlers.AddPodAdmitHandler(klet.containerManager.GetAllocateResourcesPodAdmitHandler())
```
Is it guaranteed that the AdmitHandler is called on every occasion, for example on a Kubelet restart? I'm thinking if it's ever possible to have no allocated resources when a container is added.

klueska replied:
Yes. Admit handler is always called -- even across a kubelet restart.

klueska replied:
I agree though, we will need to make sure (especially in the CPUManager), that we handle the fact that Allocate() can be called multiple times for the same container.

klueska replied:
@nolancon ^^

nolancon replied:
This seems to be already handled:  https://github.com/kubernetes/kubernetes/blob/aa964c905576d3e873994de156999f609ae8be2e/pkg/kubelet/cm/cpumanager/policy_static.go#L196-L199

Commit:
CPU Manager - Add check to policy.Allocate() for init conatiners
If container allocated CPUs is an init container, release those CPUs
back into the shared pool for re-allocation to next container.

Commit:
CPU Manager - Updates to unit tests:
- Where previously we called manager.AddContainer(), we now call both
manager.Allocate() and manager.AddContainer().
- Some test cases now have two expected errors. One each
from Allocate() and AddContainer(). Existing outcomes are unchanged.

Commit:
Device Manager - Update unit tests
- Pass container to Allocate().
- Loop through containers to call Allocate() on container by container
basis.

Commit:
Device Manager - Refactor allocatePodResources
- allocatePodResources logic altered to allow for container by container
device allocation.
- New type PodReusableDevices
- New field in devicemanager devicesToReuse

klueska committed:
Rename GetTopologyPodAmitHandler() as
GetAllocateResourcesPodAdmitHandler(). It is named as such to reflect its
new function. Also remove the Topology Manager feature gate check at higher level
kubelet.go, as it is now done in GetAllocateResourcesPodAdmitHandler().

derekwaynecarr commented:
/test pull-kubernetes-e2e-gce

derekwaynecarr commented:
/assign

nolancon commented:
/test pull-kubernetes-e2e-kind-ipv6

derekwaynecarr commented:
per sig node call, this was not wip, so updating.

derekwaynecarr commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -209,6 +209,17 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
+
+		// Check if the container that has just been allocated resources is an init container.
+		// If so, release its CPUs back into the shared pool so they can be reallocated.
+		for _, initContainer := range pod.Spec.InitContainers {
```
this makes sense per the prior commit to ensure that init containers have cpusets that are exclusive only to life of the init container.

derekwaynecarr replied:
each handler runs in sequence:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1793

all pod admission is handled in HandlePodAdditions call
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2070

the sync loop iteration is not asynch which is where all pod additions are called out from:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1894

derekwaynecarr commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -114,6 +118,9 @@ type endpointInfo struct {
 type sourcesReadyStub struct{}
 
+// PodReusableDevices is a map by pod name of devices to reuse.
+type PodReusableDevices map[string]map[string]sets.String
```
i am missing how this is bounded and if it needs mutex

derekwaynecarr commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -350,20 +358,29 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
-		if err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
-			return err
+func (m *ManagerImpl) allocatePodResources(pod *v1.Pod, container *v1.Container) error {
+	if _, ok := m.devicesToReuse[string(pod.UID)]; !ok {
```
dont we need a defer in here to handle an error case to ensure the remove happens.

derekwaynecarr commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -350,20 +358,29 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
+	if err := m.allocateContainerResources(pod, container, m.devicesToReuse[string(pod.UID)]); err != nil {
+		return err
+	}
+	m.podDevices.removeContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
```
so if we were to keep error earlier, would this map grow unbounded?  i am just trying to make sure there isnt an unbounded growth of devices to reuse.

klueska replied:
I haven’t reviewed this in detail yet, but the idea was to simply take the local devicesToReuse variable in allocatePodResources and make it a type variable on the overall manager. Once this is done, the loop in the current allocatePodResources function should be unrolled and split across a loop in the TopologManager which Allocate() on each container. The main logic from allocatePodDevices should then be pushed directly into Allocate(). If any mutetest were necessary before, they would still be necessary now.

For reference, the removal happens asynchronous here https://github.com/kubernetes/kubernetes/blob/84dfce3fed075f83b8932776ca6009f60f92979a/pkg/kubelet/cm/devicemanager/manager.go#L658, lazily cleaning up the map for terminated pods, just before allocating devices to new pods.

nolancon replied:
My takeaway from this discussion is that the only change I need make is to move logic from `allocatePodResources` back into `Allocate()` as it is now unnecessary to have the additional function.

Not mentioned here, but I will also remove the `TODO:` comment describing the new flow: https://github.com/kubernetes/kubernetes/blob/b5b395364cfdc4ea984d4f966a58fdd53edb82d5/pkg/kubelet/cm/devicemanager/manager.go#L390-L408 
I had just left this in for reviewing purposes.

Otherwise, I will leave the PR open for @klueska to give a final review

derekwaynecarr replied:
@klueska thanks for the clarification.

derekwaynecarr reviewed (changes requested):
I have a question on what controls unbounded growth of devices to reuse, but this is looking really good.

If we can clarify that concern, then this is good to go.

nolancon commented:
/test pull-kubernetes-e2e-gce

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -860,8 +874,8 @@ func (m *ManagerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Co
+		klog.V(2).Infof("needs re-allocate device plugin resources for pod %s, container %s", podUID, container.Name)
+		if err := m.Allocate(pod, container); err != nil {
 			return nil, err
 		}
```
It's unclear to me how this will interact with the need for aligned resources across all hint providers. When exactly is this `GetDeviceRunContainerOptions()` function called, and when might it attempt this reallocation of devices?

nolancon commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -672,11 +672,51 @@ func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Containe
+	// inside this component).
+	return &resourceAllocator{cm.cpuManager, cm.deviceManager}
+}
+
```
I've now added the change to `kubelet.go`, I had previously removed this commit and given it its own PR. But keeping it here means we can merge everything together and finally close the related issue https://github.com/kubernetes/kubernetes/issues/83476.

The commit (https://github.com/kubernetes/kubernetes/pull/87759/commits/850be97e347dc19b31b2048c548125d4e59bcac6) with the changes is quite small and is cherry-picked on top of existing commits so as not to dig up what has already been reviewed. 
This commit now means that this PR needs the approval of @derekwaynecarr.

nolancon commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -350,32 +358,38 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
+				return err
+			}
+			m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
+			return nil
```
Comment added.

nolancon commented:
/retest

nolancon commented:
/test pull-kubernetes-kubemark-e2e-gce-big

nolancon commented:
/test pull-kubernetes-kubemark-e2e-gce-big

nolancon commented:
/test pull-kubernetes-node-e2e-containerd

nolancon commented:
/test pull-kubernetes-node-e2e-containerd

nolancon commented:
/retest

nolancon commented:
/retest

bg-chun commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -672,11 +672,51 @@ func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Containe
+	deviceManager devicemanager.Manager
+}
+
+func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
```
It seems this admit handler is called out after topology manager calculates topology hint across all hint provider, right?

bg-chun commented on pkg/kubelet/cm/topologymanager/topology_manager.go:
```diff
@@ -227,11 +240,22 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
+		}
+		m.podTopologyHints[string(pod.UID)][container.Name] = result
+
+		err := m.allocateAlignedResources(pod, &container)
```
It seems that topology manager callout resource managers to pre-allocate compute resources here.
But I cannot find a logic to free pre-allocated resources when one of containers got error from `m.calculateAffinity()`. in this PR.
I think we should free pre-reserved resources in that case.
(Maybe it is handled by some logic... like `RemoveContainer`)

So, I`m curious that what will happen on the below situation.
-4 containers in a Pod.
-The first contaienr is admitted and resources are allocated(by `allocateAlignedResources`).
-The Second container got error from `m.calculateAffinity` and admission is rejected.

nolancon replied:
It is not really correct to say "the first container is admitted". 

The split of the original `Allocate` into two parts: `Allocate` and `AddContainer` means that in reality the "pre-allocation" is only acted upon when the pod is admitted. 

In your scenario, the pod will not be admitted due to the second container's `m.calculateAffinity` error.
Therefore the resources which were "pre-allocated" to the first container will not be consumed and those resources remain free.

bg-chun replied:
`It is not really correct to say "the first container is admitted".`
`The split of the original AddContainer into two parts`
=> I understood these, points.

`in reality the "pre-allocation" is only acted upon when the pod is admitted.`
=> But I didnt understand this point, because it seems CPU Manager(static policy) allocates cpus for container on `Allocate` part right after hint calculation of a container, so I thought cpus will be allocated even Pod is not admitted.

When the returned value of admit in below code is ture, Topology Manager calls out `Allocate` funtion of resource managers by `allocateAlignedResources()`.
`result, admit := m.calculateAffinity(pod, &container)`

Here, in the case of CPU Manager, CPUManager calls out `staticPolicy.Allocate(..)`.
(I looked only CPU Manager.)

Then It seems static policy : 
1) allocates cpus based on stored affinity for container by `staticPolicy.allocateCPUs() and takeByTopology() internally`
2) update `state` of cpus to exclude allocated cpus by the below code.
    `s.SetDefaultCPUSet(s.GetDefaultCPUSet().Difference(result))`

klueska replied:
It's possible that CPUs may get allocated even if the pod is ultimately not admitted. However, any cpus allocated to non-admitted pods (as well as any pods that may have terminated for some other reason unbeknownst to the kubelet) are lazily reclaimed at the top of `GetTopologyHints()` here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpumanager/cpu_manager.go#L303

bg-chun replied:
Oh, I missed that garbage collector, I will take a look it.
Thankyou @nolancon @klueska

nolancon replied:
I've tested out your scenario locally @bg-chun and it works as expected in @klueska's description.

derekwaynecarr commented:
/hold cancel

the updates look good.
i added details on pod admission sequence for how it is serialized.

/approve
/lgtm

nolancon commented:
/retest

liggitt commented:
This broke clusters brought up with beta features disabled. See conformance test at https://testgrid.k8s.io/conformance-kind#conformance-ga-only

```
Mar 05 09:11:48 kind-control-plane kubelet[334]: E0305 09:11:48.647899     334 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
Mar 05 09:11:48 kind-control-plane kubelet[334]: goroutine 218 [running]:
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3a7d7c0, 0x6e8bf20)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
Mar 05 09:11:48 kind-control-plane kubelet[334]: panic(0x3a7d7c0, 0x6e8bf20)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         GOROOT/src/runtime/panic.go:679 +0x1b2
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet/cm.(*resourceAllocator).Admit(0xc000886940, 0xc000f49180, 0x0, 0x0, 0x0, 0x0, 0x0)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/cm/container_manager_linux.go:709 +0x1fb
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).canAdmitPod(0xc000cca000, 0x0, 0x0, 0x0, 0xc000914c00, 0x0, 0x0, 0x0, 0x2, 0xc00100a160)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/kubelet.go:1793 +0xea
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).HandlePodAdditions(0xc000cca000, 0xc00087c360, 0x4, 0x4)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/kubelet.go:2074 +0x30b
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoopIteration(0xc000cca000, 0xc000b1f320, 0x4b45680, 0xc000cca000, 0xc000f4be00, 0xc000f4be60, 0xc000a96ea0, 0x3bebb80)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/kubelet.go:1912 +0x55a
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop(0xc000cca000, 0xc000b1f320, 0x4b45680, 0xc000cca000)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/kubelet.go:1855 +0x371
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run(0xc000cca000, 0xc000b1f320)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         pkg/kubelet/kubelet.go:1452 +0x27c
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/cmd/kubelet/app.startKubelet.func1()
Mar 05 09:11:48 kind-control-plane kubelet[334]:         cmd/kubelet/app/server.go:1133 +0x3c
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000886d60)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5e
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000886d60, 0x4a8ff00, 0xc000884e10, 0xc000116801, 0xc0000c40c0)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xa3
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000886d60, 0x0, 0x0, 0x1, 0xc0000c40c0)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0xe2
Mar 05 09:11:48 kind-control-plane kubelet[334]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc000886d60, 0x0, 0xc0000c40c0)
Mar 05 09:11:48 kind-control-plane kubelet[334]:         staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
Mar 05 09:11:48 kind-control-plane kubelet[334]: created by k8s.io/kubernetes/cmd/kubelet/app.startKubelet
Mar 05 09:11:48 kind-control-plane kubelet[334]:         cmd/kubelet/app/server.go:1132 +0x9a
```

liggitt commented:
Last good run was d682c8389
First bad run was 364c66ed9

git log --format=oneline d682c8389..364c66ed9 -- pkg/kubelet/cm
ac32644d6e712af0747f39a945c8df29d6a9cfc4 Merge pull request #87759 from klueska/upstream-move-cpu-allocation-to-pod-admit
2327934a8602d2f0dad369a86c5475d27fc1b062 Rename GetTopologyPodAmitHandler() as GetAllocateResourcesPodAdmitHandler(). It is named as such to reflect its new function. Also remove the Topology Manager feature gate check at higher level kubelet.go, as it is now done in GetAllocateResourcesPodAdmitHandler().
a9c61295779551c85b86099a437c7ea556f857a2 Device Manager - Update unit tests - Pass container to Allocate(). - Loop through containers to call Allocate() on container by container basis.
cb9fdc49db8ed56aea6569e1a9ab69920ae8cc2e Device Manager - Refactor allocatePodResources - allocatePodResources logic altered to allow for container by container device allocation. - New type PodReusableDevices - New field in devicemanager devicesToReuse
0a9bd0334dd315d73619fa8830750f5113b296e6 CPU Manager - Updates to unit tests: - Where previously we called manager.AddContainer(), we now call both manager.Allocate() and manager.AddContainer(). - Some test cases now have two expected errors. One each from Allocate() and AddContainer(). Existing outcomes are unchanged.
467f66580b970c727a51d361f4dba82b7bd0c9c2 CPU Manager - Add check to policy.Allocate() for init conatiners If container allocated CPUs is an init container, release those CPUs back into the shared pool for re-allocation to next container.
709989efa2a039007c78ae1754fa4eb4a5fb5d79 CPU Manager - Rename policy.AddContainer() to policy.Allocate()
0d68bffd03f1af15fc8cbb3eaeda9ff95a47eeab Change GetTopologyPodAdmitHandler() to be more general
0b168f0243660a4b5f5c65d2b76ce828362ffa88 Change devicemanager to implement HintProvider.Allocate()
91f91858a54bd1c5221026fab8fe7539c2c401a5 Change CPUManager to implement HintProvider.Allocate()
9e4ee5ecc3f0c680632d7f7f6287c3da86d757d3 Add Allocate() call to TopologyManager's HintProvider interface
a3f099ea4d7c759b9cd014aec8e3011b2363efde Split devicemanager Allocate into two functions

vpickard commented:
I'm taking a look at this now, and have also reached out to Conor. Will
update shortly....

On Thu, Mar 5, 2020 at 10:14 AM Jordan Liggitt <notifications@github.com>
wrote:

> Last good run was d682c83
> <https://github.com/kubernetes/kubernetes/commit/d682c8389087e55835bbd02ce20cb6110c81731f>
> First bad run was 364c66e
> <https://github.com/kubernetes/kubernetes/commit/364c66ed9c7125e55ced73b35f00a9a73fd79e96>
>
> git log --format=oneline d682c83
> <https://github.com/kubernetes/kubernetes/commit/d682c8389087e55835bbd02ce20cb6110c81731f>
> ..364c66e
> <https://github.com/kubernetes/kubernetes/commit/364c66ed9c7125e55ced73b35f00a9a73fd79e96>
> -- pkg/kubelet/cm
> ac32644
> <https://github.com/kubernetes/kubernetes/commit/ac32644d6e712af0747f39a945c8df29d6a9cfc4>
> Merge pull request #87759
> <https://github.com/kubernetes/kubernetes/pull/87759> from
> klueska/upstream-move-cpu-allocation-to-pod-admit
> 2327934
> <https://github.com/kubernetes/kubernetes/commit/2327934a8602d2f0dad369a86c5475d27fc1b062>
> Rename GetTopologyPodAmitHandler() as
> GetAllocateResourcesPodAdmitHandler(). It is named as such to reflect its
> new function. Also remove the Topology Manager feature gate check at higher
> level kubelet.go, as it is now done in
> GetAllocateResourcesPodAdmitHandler().
> a9c6129
> <https://github.com/kubernetes/kubernetes/commit/a9c61295779551c85b86099a437c7ea556f857a2>
> Device Manager - Update unit tests - Pass container to Allocate(). - Loop
> through containers to call Allocate() on container by container basis.
> cb9fdc4
> <https://github.com/kubernetes/kubernetes/commit/cb9fdc49db8ed56aea6569e1a9ab69920ae8cc2e>
> Device Manager - Refactor allocatePodResources - allocatePodResources logic
> altered to allow for container by container device allocation. - New type
> PodReusableDevices - New field in devicemanager devicesToReuse
> 0a9bd03
> <https://github.com/kubernetes/kubernetes/commit/0a9bd0334dd315d73619fa8830750f5113b296e6>
> CPU Manager - Updates to unit tests: - Where previously we called
> manager.AddContainer(), we now call both manager.Allocate() and
> manager.AddContainer(). - Some test cases now have two expected errors. One
> each from Allocate() and AddContainer(). Existing outcomes are unchanged.
> 467f665
> <https://github.com/kubernetes/kubernetes/commit/467f66580b970c727a51d361f4dba82b7bd0c9c2>
> CPU Manager - Add check to policy.Allocate() for init conatiners If
> container allocated CPUs is an init container, release those CPUs back into
> the shared pool for re-allocation to next container.
> 709989e
> <https://github.com/kubernetes/kubernetes/commit/709989efa2a039007c78ae1754fa4eb4a5fb5d79>
> CPU Manager - Rename policy.AddContainer() to policy.Allocate()
> 0d68bff
> <https://github.com/kubernetes/kubernetes/commit/0d68bffd03f1af15fc8cbb3eaeda9ff95a47eeab>
> Change GetTopologyPodAdmitHandler() to be more general
> 0b168f0
> <https://github.com/kubernetes/kubernetes/commit/0b168f0243660a4b5f5c65d2b76ce828362ffa88>
> Change devicemanager to implement HintProvider.Allocate()
> 91f9185
> <https://github.com/kubernetes/kubernetes/commit/91f91858a54bd1c5221026fab8fe7539c2c401a5>
> Change CPUManager to implement HintProvider.Allocate()
> 9e4ee5e
> <https://github.com/kubernetes/kubernetes/commit/9e4ee5ecc3f0c680632d7f7f6287c3da86d757d3>
> Add Allocate() call to TopologyManager's HintProvider interface
> a3f099e
> <https://github.com/kubernetes/kubernetes/commit/a3f099ea4d7c759b9cd014aec8e3011b2363efde>
> Split devicemanager Allocate into two functions
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/87759?email_source=notifications&email_token=AEC5QWGK5ULIHJYRV5L3AXDRF66WTA5CNFSM4KO4QYNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN5URFY#issuecomment-595282071>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEC5QWEZ4BQ5XLIHF4WLWELRF66WTANCNFSM4KO4QYNA>
> .
>

bg-chun commented:
https://github.com/kubernetes/kubernetes/blob/1f2e1967d1ca44d5c4438a169eda83f69d82bee5/pkg/kubelet/cm/container_manager_linux.go#L688

Do we need to return `&resourceAllocator{cm.cpuManager, cm.deviceManager}` here?
I cannot see any reason.

liggitt commented:
it seems like the places that use cpuManager should check if it is non-nil (since it is optional, depending on whether the feature is enabled)

klueska commented:
If this is not done, then no allocations will be made by either component
if the TopologyManager feature gate is disabled.

Byonggon Chun <notifications@github.com> schrieb am Do. 5. März 2020 um
16:23:

>
> https://github.com/kubernetes/kubernetes/blob/1f2e1967d1ca44d5c4438a169eda83f69d82bee5/pkg/kubelet/cm/container_manager_linux.go#L688
>
> Do we need to return &resourceAllocator{cm.cpuManager, cm.deviceManager}
> here?
> I cannot see any reason.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/87759?email_source=notifications&email_token=AAAXRI7AD7YJEFDLT7ZIEGDRF67VPA5CNFSM4KO4QYNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN5VTOQ#issuecomment-595286458>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAXRI5HIYFUTA7PCK4IAFDRF67VPANCNFSM4KO4QYNA>
> .
>
-- 
~Kevin

liggitt commented:
something like this:
```
diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
index 98c3647193f..13c7176bdc2 100644
--- a/pkg/kubelet/cm/container_manager_linux.go
+++ b/pkg/kubelet/cm/container_manager_linux.go
@@ -706,12 +706,14 @@ func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle
                        }
                }
 
-               err = m.cpuManager.Allocate(pod, &container)
-               if err != nil {
-                       return lifecycle.PodAdmitResult{
-                               Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
-                               Reason:  "UnexpectedAdmissionError",
-                               Admit:   false,
+               if m.cpuManager != nil {
+                       err = m.cpuManager.Allocate(pod, &container)
+                       if err != nil {
+                               return lifecycle.PodAdmitResult{
+                                       Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
+                                       Reason:  "UnexpectedAdmissionError",
+                                       Admit:   false,
+                               }
                        }
                }
        }
diff --git a/pkg/kubelet/cm/internal_container_lifecycle.go b/pkg/kubelet/cm/internal_container_lifecycle.go
index 1176a42a8d1..9e243430269 100644
--- a/pkg/kubelet/cm/internal_container_lifecycle.go
+++ b/pkg/kubelet/cm/internal_container_lifecycle.go
@@ -38,7 +38,7 @@ type internalContainerLifecycleImpl struct {
 }
 
 func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
-       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
+       if i.cpuManager != nil {
                err := i.cpuManager.AddContainer(pod, container, containerID)
                if err != nil {
                        return err
@@ -54,14 +54,14 @@ func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, containe
 }
 
 func (i *internalContainerLifecycleImpl) PreStopContainer(containerID string) error {
-       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
+       if i.cpuManager != nil {
                return i.cpuManager.RemoveContainer(containerID)
        }
        return nil
 }
 
 func (i *internalContainerLifecycleImpl) PostStopContainer(containerID string) error {
-       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
+       if i.cpuManager != nil {
                err := i.cpuManager.RemoveContainer(containerID)
                if err != nil {
                        return err
```

bg-chun commented:
@klueska klueska 
Yeah, I missed this part.
https://github.com/kubernetes/kubernetes/blob/1f2e1967d1ca44d5c4438a169eda83f69d82bee5/pkg/kubelet/cm/container_manager_linux.go#L696

nolancon commented:
> something like this:
> 
> ```diff
> diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
> index 98c3647193f..13c7176bdc2 100644
> --- a/pkg/kubelet/cm/container_manager_linux.go
> +++ b/pkg/kubelet/cm/container_manager_linux.go
> @@ -706,12 +706,14 @@ func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle
>                         }
>                 }
>  
> -               err = m.cpuManager.Allocate(pod, &container)
> -               if err != nil {
> -                       return lifecycle.PodAdmitResult{
> -                               Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
> -                               Reason:  "UnexpectedAdmissionError",
> -                               Admit:   false,
> +               if m.cpuManager != nil {
> +                       err = m.cpuManager.Allocate(pod, &container)
> +                       if err != nil {
> +                               return lifecycle.PodAdmitResult{
> +                                       Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
> +                                       Reason:  "UnexpectedAdmissionError",
> +                                       Admit:   false,
> +                               }
>                         }
>                 }
>         }
> diff --git a/pkg/kubelet/cm/internal_container_lifecycle.go b/pkg/kubelet/cm/internal_container_lifecycle.go
> index 1176a42a8d1..9e243430269 100644
> --- a/pkg/kubelet/cm/internal_container_lifecycle.go
> +++ b/pkg/kubelet/cm/internal_container_lifecycle.go
> @@ -38,7 +38,7 @@ type internalContainerLifecycleImpl struct {
>  }
>  
>  func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
> -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> +       if i.cpuManager != nil {
>                 err := i.cpuManager.AddContainer(pod, container, containerID)
>                 if err != nil {
>                         return err
> @@ -54,14 +54,14 @@ func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, containe
>  }
>  
>  func (i *internalContainerLifecycleImpl) PreStopContainer(containerID string) error {
> -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> +       if i.cpuManager != nil {
>                 return i.cpuManager.RemoveContainer(containerID)
>         }
>         return nil
>  }
>  
>  func (i *internalContainerLifecycleImpl) PostStopContainer(containerID string) error {
> -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> +       if i.cpuManager != nil {
>                 err := i.cpuManager.RemoveContainer(containerID)
>                 if err != nil {
>                         return err
> ```

@liggitt Just looking for clarification on this - is this something we need to fix immediately? My take is #16632 makes the job optional, so we have some time to test out the fix and make sure all is ok?

liggitt commented:
Yes, it needs to be fixed immediately. Loss of signal on that job means other things that break GA-only clusters can merge without visibility. Opened https://github.com/kubernetes/kubernetes/issues/88855 to track this.

vpickard commented:
cm.CPUManager will be nil for ga-only features, so it seems like the proper fix will be something like @liggitt suggested. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager_linux.go#L318

liggitt commented:
> My take is kubernetes/test-infra#16632 makes the job optional, so we have some time to test out the fix and make sure all is ok?

The job introduced in https://github.com/kubernetes/test-infra/pull/16632 is optional only until we gain confidence in the job itself. The thing the job is testing (being able to bring up GA-only clusters) is definitely not optional.

vpickard commented:
> > something like this:
> > ```diff
> > diff --git a/pkg/kubelet/cm/container_manager_linux.go b/pkg/kubelet/cm/container_manager_linux.go
> > index 98c3647193f..13c7176bdc2 100644
> > --- a/pkg/kubelet/cm/container_manager_linux.go
> > +++ b/pkg/kubelet/cm/container_manager_linux.go
> > @@ -706,12 +706,14 @@ func (m *resourceAllocator) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle
> >                         }
> >                 }
> >  
> > -               err = m.cpuManager.Allocate(pod, &container)
> > -               if err != nil {
> > -                       return lifecycle.PodAdmitResult{
> > -                               Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
> > -                               Reason:  "UnexpectedAdmissionError",
> > -                               Admit:   false,
> > +               if m.cpuManager != nil {
> > +                       err = m.cpuManager.Allocate(pod, &container)
> > +                       if err != nil {
> > +                               return lifecycle.PodAdmitResult{
> > +                                       Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
> > +                                       Reason:  "UnexpectedAdmissionError",
> > +                                       Admit:   false,
> > +                               }
> >                         }
> >                 }
> >         }
> > diff --git a/pkg/kubelet/cm/internal_container_lifecycle.go b/pkg/kubelet/cm/internal_container_lifecycle.go
> > index 1176a42a8d1..9e243430269 100644
> > --- a/pkg/kubelet/cm/internal_container_lifecycle.go
> > +++ b/pkg/kubelet/cm/internal_container_lifecycle.go
> > @@ -38,7 +38,7 @@ type internalContainerLifecycleImpl struct {
> >  }
> >  
> >  func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
> > -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> > +       if i.cpuManager != nil {
> >                 err := i.cpuManager.AddContainer(pod, container, containerID)
> >                 if err != nil {
> >                         return err
> > @@ -54,14 +54,14 @@ func (i *internalContainerLifecycleImpl) PreStartContainer(pod *v1.Pod, containe
> >  }
> >  
> >  func (i *internalContainerLifecycleImpl) PreStopContainer(containerID string) error {
> > -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> > +       if i.cpuManager != nil {
> >                 return i.cpuManager.RemoveContainer(containerID)
> >         }
> >         return nil
> >  }
> >  
> >  func (i *internalContainerLifecycleImpl) PostStopContainer(containerID string) error {
> > -       if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUManager) {
> > +       if i.cpuManager != nil {
> >                 err := i.cpuManager.RemoveContainer(containerID)
> >                 if err != nil {
> >                         return err
> > ```
> 
> @liggitt Just looking for clarification on this - is this something we need to fix immediately? My take is #16632 makes the job optional, so we have some time to test out the fix and make sure all is ok?

I think a similiar nil check would be needed here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager_linux.go#L688

liggitt commented:
I don't think so, as long as resourceAllocator handles a nil cpuManager correctly

vpickard commented:
> I don't think so, as long as resourceAllocator handles a nil cpuManager correctly

Hm, I don't see resourceAllocator handling a nil cpuManager. The resourceAllocator will just be added to the list of handlers.... with nil cpuManager... unless I missed something. 

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/interfaces.go#L104

Which makes me wonder, what should be returned if there is no cpumanager because the feature gate is effectively disabled (since cpumanager is not GA)?

klueska commented:
I agree. @liggitt's diff should be enough to fix this.
@vpickard or @nolancon Can you put a PR together or do you want me to?

nolancon commented:
> I agree. @liggitt's diff should be enough to fix this.
> @vpickard or @nolancon Can you put a PR together or do you want me to?

https://github.com/kubernetes/kubernetes/pull/88857

klueska commented:
> > I agree. @liggitt's diff should be enough to fix this.
> > @vpickard or @nolancon Can you put a PR together or do you want me to?
> 
> #88857

Sorry, I missed that. My page wasn't refreshed. I have added an approve on it.
@liggitt can you give it an /lgtm if you think it looks good.
